browser_console.js

```diff
@@ -0,0 +1,159 @@
+/**
+ * Reconstruct original types from serialized message
+ * @param {*} value
+ */
+function parseConsoleArg(value) {
+    if (Array.isArray(value)) {
+        return value.map(item => parseConsoleArg(item));
+    } else if (typeof value === 'object' && value !== null) {
+        if (value.__pentf_serialized) {
+            if (value.type === 'undefined') {
+                return undefined;
+            } else if (value.type === 'Error') {
+                const err = new Error(value.message);
+                err.stack = value.stack;
+                return err;
+            } else if (value.type === 'Set') {
+                return new Set(value.items.map(item => parseConsoleArg(item)));
+            } else if (value.type === 'Map') {
+                return new Map(
+                    value.items.map(item => {
+                        return [parseConsoleArg(item[0]), parseConsoleArg(item[1])];
+                    })
+                );
+            } else if (value.type === 'Function') {
+                // Mirror node output format
+                return !value.name ? '[Function (anonymous)]' : `[Function: ${value.name}]`;
+            }
+        }
+
+        let out = {};
+        for (const key in value) {
+            out[key] = parseConsoleArg(value[key]);
+        }
+        return out;
+    }
+
+    return value;
+}
+
+/**
+ * Serialize any JS value to JSON. Used to send data from the browser
+ * to the node process.
+ * @param {*} value
+ * @param {Set<any>} seen
+ */
+function serialize(value, seen) {
+    if (seen.has(value)) {
+        return '[[Circular]]';
+    }
+
+    if (Array.isArray(value)) {
+        seen.add(value);
+        return value.map(x => serialize(x, seen));
+    } else if (value === undefined) {
+        return {
+            __pentf_serialized: true,
+            type: 'undefined',
+        };
+    } else if (value === null) {
+        return null;
+    } else if (typeof value === 'object') {
+        if (value instanceof Error) {
```

Comment:
Why is it correct not to add the object here to `seen`? For example, a Map could contain itself. Errors and Sets shouldn't usually contain themselves, but it's technically possible.

Reply:
Good point! Was an oversight on my end.